### PR TITLE
Add ability to reorder fields in DatagridConfigurable and SimpleFormConfigurable

### PR DESCRIPTION
**Pull request by fzaninotto**
https://user-images.githubusercontent.com/99944/203121514-5af507a4-1ce7-42f9-b95b-573f81ca16b9.mov

https://user-images.githubusercontent.com/99944/203122116-6ea76344-bcdf-4b79-923a-10452f11b7f4.mov

### CODE REVIEW

**Review by WiXSL — APPROVED** [review]

**Review by WiXSL — CHANGES_REQUESTED** [review]
I believe you forgot to update the screencast.
And maybe a few tests wouldn't hurt 😃

**Comment by fzaninotto** [comment]
Screencasts added. As for the tests, they are left as an exercise for the reader ;)

File: packages/ra-ui-materialui/src/preferences/FieldsSelector.tsx [hunk 1]
```diff
@@ -43,14 +42,52 @@ export const FieldsSelector = ({
...
+            ];
+        }
+        setAvailableFields(newAvailableFields);
+        setFields(fields =>
+            newAvailableFields
+                .filter(field => !omit?.includes(field.source))
+                .filter(field => fields.includes(field.index))
+                .map(field => field.index)
+        );
+    };
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/preferences/FieldsSelector.tsx:80]: Fun fact: this method was almost entirely written by gitHub Copilot!

**Review by WiXSL — APPROVED** [review]

File: packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.tsx [hunk 2]
```diff
@@ -86,6 +85,43 @@ export const SelectColumnsButton = props => {
...
+                ...availableColumns.slice(0, index1Pos),
+                ...availableColumns.slice(index1Pos + 1, index2Pos + 1),
+                availableColumns[index1Pos],
+                ...availableColumns.slice(index2Pos + 1),
+            ];
+        }
+        setAvailableColumns(newAvailableColumns);
+        setColumns(columns =>
+            newAvailableColumns
+                .filter(column => !omit?.includes(column.source))
```
Comment (slax57) [inline packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.tsx:119]: I don't think you should use `omit` here. Reordering the columns shouldn't re-apply the omit param IMO.

File: packages/ra-ui-materialui/src/preferences/FieldsSelector.tsx [hunk 3]
```diff
@@ -43,14 +42,52 @@ export const FieldsSelector = ({
...
+                ...availableFields.slice(0, index1Pos),
+                ...availableFields.slice(index1Pos + 1, index2Pos + 1),
+                availableFields[index1Pos],
+                ...availableFields.slice(index2Pos + 1),
+            ];
+        }
+        setAvailableFields(newAvailableFields);
+        setFields(fields =>
+            newAvailableFields
+                .filter(field => !omit?.includes(field.source))
```
Comment (slax57) [inline packages/ra-ui-materialui/src/preferences/FieldsSelector.tsx:76]: same

**Review by slax57 — CHANGES_REQUESTED** [review]

**Review by slax57 — APPROVED** [review]

**Pull request merged**